File: erasure/format.go
package erasure

import (
	"encoding/json"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
)

// DiskFormat represents the format.json structure from MinIO
type DiskFormat struct {
	Version string `json:"version"`
	Format  string `json:"format"`
	ID      string `json:"id"` // Pool ID
	XL      struct {
		Version          string     `json:"version"`
		This             string     `json:"this"`
		Sets             [][]string `json:"sets"`
		DistributionAlgo string     `json:"distributionAlgo"`
	} `json:"xl"`
}

// DiskInfo contains information about a discovered disk
type DiskInfo struct {
	Path      string // Filesystem path to the disk
	UUID      string // Disk UUID from format.json
	PoolIndex int    // Which pool (0, 1, ...)
	SetIndex  int    // Which erasure set within the pool
	DiskIndex int    // Position within the erasure set (0-based)
	PoolID    string // Pool UUID
}

// PoolConfig contains the discovered pool configuration
type PoolConfig struct {
	PoolID    string       // Pool UUID
	PoolIndex int          // Index of this pool (0, 1, ...)
	Sets      [][]DiskInfo // Disks organized by set
	SetCount  int          // Number of erasure sets
}

// ClusterConfig contains all discovered pools
type ClusterConfig struct {
	Pools []*PoolConfig
}

// TotalSets returns total number of sets across all pools
func (c *ClusterConfig) TotalSets() int {
	total := 0
	for _, pool := range c.Pools {
		total += pool.SetCount
	}
	return total
}

// DiscoverCluster reads format.json from all disks and groups them by pool
func DiscoverCluster(diskPaths []string) (*ClusterConfig, error) {
	if len(diskPaths) == 0 {
		return nil, fmt.Errorf("no disk paths provided")
	}

	// Read format.json from each disk
	type diskData struct {
		format DiskFormat
		path   string
	}
	var disks []diskData

	for _, path := range diskPaths {
		formatPath := filepath.Join(path, ".minio.sys", "format.json")
		data, err := os.ReadFile(formatPath)
		if err != nil {
			continue
		}

		var format DiskFormat
		if err := json.Unmarshal(data, &format); err != nil {
			continue
		}

		disks = append(disks, diskData{format: format, path: path})
	}

	if len(disks) == 0 {
		return nil, fmt.Errorf("no valid format.json found in any disk")
	}

	// Group disks by pool ID
	poolDisks := make(map[string][]diskData)
	poolOrder := []string{} // Preserve discovery order

	for _, d := range disks {
		poolID := d.format.ID
		if _, exists := poolDisks[poolID]; !exists {
			poolOrder = append(poolOrder, poolID)
		}
		poolDisks[poolID] = append(poolDisks[poolID], d)
	}

	// Build cluster config
	cluster := &ClusterConfig{
		Pools: make([]*PoolConfig, len(poolOrder)),
	}

	for poolIdx, poolID := range poolOrder {
		disksInPool := poolDisks[poolID]

		// Get sets configuration from first disk in pool
		sets := disksInPool[0].format.XL.Sets
		if len(sets) == 0 {
			return nil, fmt.Errorf("pool %s has no erasure sets", poolID)
		}

		// Build UUID to path mapping for this pool
		uuidToPath := make(map[string]string)
		for _, d := range disksInPool {
			uuidToPath[d.format.XL.This] = d.path
		}

		// Build pool config
		pool := &PoolConfig{
			PoolID:    poolID,
			PoolIndex: poolIdx,
			SetCount:  len(sets),
			Sets:      make([][]DiskInfo, len(sets)),
		}

		for setIdx, set := range sets {
			pool.Sets[setIdx] = make([]DiskInfo, len(set))
			for diskIdx, uuid := range set {
				path := uuidToPath[uuid]
				pool.Sets[setIdx][diskIdx] = DiskInfo{
					Path:      path,
					UUID:      uuid,
					PoolIndex: poolIdx,
					SetIndex:  setIdx,
					DiskIndex: diskIdx,
					PoolID:    poolID,
				}
			}
		}

		cluster.Pools[poolIdx] = pool
	}

	return cluster, nil
}

// DiscoverClusterInDirectory finds all storage directories and returns cluster configuration
func DiscoverClusterInDirectory(rootDir string) (*ClusterConfig, error) {
	entries, err := os.ReadDir(rootDir)
	if err != nil {
		return nil, fmt.Errorf("read root directory: %w", err)
	}

	var diskPaths []string
	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}

		path := filepath.Join(rootDir, entry.Name())
		formatPath := filepath.Join(path, ".minio.sys", "format.json")
		if _, err := os.Stat(formatPath); err == nil {
			diskPaths = append(diskPaths, path)
		}
	}

	if len(diskPaths) == 0 {
		return nil, fmt.Errorf("no MinIO disks found in %s", rootDir)
	}

	return DiscoverCluster(diskPaths)
}

// Legacy functions for backwards compatibility

// DiscoverPool returns the first pool only (deprecated, use DiscoverCluster)
func DiscoverPool(diskPaths []string) (*PoolConfig, error) {
	cluster, err := DiscoverCluster(diskPaths)
	if err != nil {
		return nil, err
	}
	if len(cluster.Pools) == 0 {
		return nil, fmt.Errorf("no pools found")
	}
	return cluster.Pools[0], nil
}

// DiscoverPoolInDirectory returns the first pool only (deprecated)
func DiscoverPoolInDirectory(rootDir string) (*PoolConfig, error) {
	cluster, err := DiscoverClusterInDirectory(rootDir)
	if err != nil {
		return nil, err
	}
	if len(cluster.Pools) == 0 {
		return nil, fmt.Errorf("no pools found")
	}
	return cluster.Pools[0], nil
}

// DiscoverDisks returns the first set of the first pool (deprecated)
func DiscoverDisks(diskPaths []string) ([]DiskInfo, error) {
	pool, err := DiscoverPool(diskPaths)
	if err != nil {
		return nil, err
	}
	if len(pool.Sets) == 0 {
		return nil, fmt.Errorf("no sets found")
	}
	return pool.Sets[0], nil
}

// DiscoverDisksInDirectory returns the first set of the first pool (deprecated)
func DiscoverDisksInDirectory(rootDir string) ([]DiskInfo, error) {
	pool, err := DiscoverPoolInDirectory(rootDir)
	if err != nil {
		return nil, err
	}
	if len(pool.Sets) == 0 {
		return nil, fmt.Errorf("no sets found")
	}
	return pool.Sets[0], nil
}

// GetOrderedDiskPaths returns disk paths from a slice of DiskInfo
func GetOrderedDiskPaths(diskInfos []DiskInfo) []string {
	paths := make([]string, len(diskInfos))
	for i, info := range diskInfos {
		paths[i] = info.Path
	}
	return paths
}

// DiskMapping maps disk UUIDs to partition indices for raw disk access
type DiskMapping struct {
	UUIDToPartition map[string]int // disk UUID -> partition index
	PartitionToUUID map[int]string // partition index -> disk UUID
	DiskOrder       []string       // ordered list of disk UUIDs from format.json
	PoolID          string         // pool ID
}

// GetPartitionForDiskIndex returns the partition index for a given disk index in the distribution
func (dm *DiskMapping) GetPartitionForDiskIndex(diskIndex int) (int, bool) {
	if diskIndex < 0 || diskIndex >= len(dm.DiskOrder) {
		return 0, false
	}
	uuid := dm.DiskOrder[diskIndex]
	partIdx, ok := dm.UUIDToPartition[uuid]
	return partIdx, ok
}

// BuildDiskMappingFromRaw reads format.json from raw partitions and builds the disk mapping
func BuildDiskMappingFromRaw(partitions []*RawFS) (*DiskMapping, error) {
	if len(partitions) == 0 {
		return nil, fmt.Errorf("no partitions provided")
	}

	dm := &DiskMapping{
		UUIDToPartition: make(map[string]int),
		PartitionToUUID: make(map[int]string),
	}

	// Read format.json from each partition
	for partIdx, part := range partitions {
		if part == nil {
			continue // Skip failed partitions
		}
		format, err := readFormatFromFS(part.FS())
		if err != nil {
			// Skip partitions without format.json
			continue
		}

		// Store this partition's UUID
		diskUUID := format.XL.This
		dm.UUIDToPartition[diskUUID] = partIdx
		dm.PartitionToUUID[partIdx] = diskUUID

		// Use the first partition's format.json to get the disk order
		if dm.DiskOrder == nil && len(format.XL.Sets) > 0 {
			dm.DiskOrder = format.XL.Sets[0]
			dm.PoolID = format.ID
		}
	}

	if dm.DiskOrder == nil {
		return nil, fmt.Errorf("no valid format.json found in any partition")
	}

	return dm, nil
}

// readFormatFromFS reads format.json from an fs.FS
func readFormatFromFS(fsys fs.FS) (*DiskFormat, error) {
	f, err := fsys.Open(".minio.sys/format.json")
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var format DiskFormat
	if err := json.NewDecoder(f).Decode(&format); err != nil {
		return nil, err
	}
	return &format, nil
}

File: minio-unfuck-rs/crates/mfu-core/src/format.rs
//! format.json types and cluster discovery
//!
//! Port of erasure/format.go. Parses MinIO's format.json to discover cluster topology.

use std::collections::HashMap;

use anyhow::{bail, Result};
use serde::Deserialize;

use crate::types::{ClusterConfig, DiskInfo, PoolConfig};

/// Raw format.json structure from MinIO
#[derive(Debug, Clone, Deserialize)]
pub struct DiskFormat {
    pub version: String,
    pub format: String,
    pub id: String, // Pool ID
    pub xl: XLFormat,
}

#[derive(Debug, Clone, Deserialize)]
pub struct XLFormat {
    pub version: String,
    pub this: String, // This disk's UUID
    pub sets: Vec<Vec<String>>,
    #[serde(rename = "distributionAlgo")]
    pub distribution_algo: String,
}

/// Parse a format.json from bytes
pub fn parse_format(data: &[u8]) -> Result<DiskFormat> {
    let format: DiskFormat = serde_json::from_slice(data)?;
    Ok(format)
}

/// Build cluster config from multiple format.json files.
///
/// `formats` is a list of (device_id, DiskFormat) pairs.
pub fn build_cluster_config(formats: &[(usize, DiskFormat)]) -> Result<ClusterConfig> {
    if formats.is_empty() {
        bail!("no format.json files provided");
    }

    // Group by pool ID, preserving discovery order
    let mut pool_disks: HashMap<String, Vec<(usize, &DiskFormat)>> = HashMap::new();
    let mut pool_order: Vec<String> = Vec::new();

    for (device_id, fmt) in formats {
        let pool_id = &fmt.id;
        if !pool_disks.contains_key(pool_id) {
            pool_order.push(pool_id.clone());
        }
        pool_disks
            .entry(pool_id.clone())
            .or_default()
            .push((*device_id, fmt));
    }

    // Build cluster config
    let mut pools = Vec::with_capacity(pool_order.len());

    for (pool_idx, pool_id) in pool_order.iter().enumerate() {
        let disks_in_pool = &pool_disks[pool_id];

        // Get sets configuration from first disk in pool
        let sets_config = &disks_in_pool[0].1.xl.sets;
        if sets_config.is_empty() {
            bail!("pool {} has no erasure sets", pool_id);
        }

        // Build UUID to device_id mapping
        let mut uuid_to_device: HashMap<String, usize> = HashMap::new();
        for (device_id, fmt) in disks_in_pool {
            uuid_to_device.insert(fmt.xl.this.clone(), *device_id);
        }

        // Build pool config
        let mut sets = Vec::with_capacity(sets_config.len());

        for (set_idx, set) in sets_config.iter().enumerate() {
            let mut disk_infos = Vec::with_capacity(set.len());
            for (disk_idx, uuid) in set.iter().enumerate() {
                let device_id = uuid_to_device.get(uuid).copied();
                disk_infos.push(DiskInfo {
                    uuid: uuid.clone(),
                    pool_index: pool_idx,
                    set_index: set_idx,
                    disk_index: disk_idx,
                    pool_id: pool_id.clone(),
                    device_id,
                });
            }
            sets.push(disk_infos);
        }

        pools.push(PoolConfig {
            pool_id: pool_id.clone(),
            pool_index: pool_idx,
            sets,
        });
    }

    Ok(ClusterConfig { pools })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    fn fixture_path(name: &str) -> PathBuf {
        let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        path.pop();
        path.pop();
        path.push(".disks");
        path.push(name);
        path
    }

    #[test]
    fn test_parse_format_json() {
        let path = fixture_path("storage1/.minio.sys/format.json");
        if !path.exists() {
            eprintln!("skipping test: fixture not found");
            return;
        }
        let data = std::fs::read(&path).unwrap();
        let fmt = parse_format(&data).unwrap();

        assert_eq!(fmt.version, "1");
        assert_eq!(fmt.format, "xl");
        assert!(!fmt.id.is_empty());
        assert!(!fmt.xl.this.is_empty());
        assert!(!fmt.xl.sets.is_empty());
        assert_eq!(fmt.xl.sets[0].len(), 16);
    }

    #[test]
    fn test_build_cluster_config() {
        let mut formats = Vec::new();
        for i in 1..=16 {
            let path = fixture_path(&format!("storage{}/.minio.sys/format.json", i));
            if !path.exists() {
                eprintln!("skipping test: fixtures not found");
                return;
            }
            let data = std::fs::read(&path).unwrap();
            let fmt = parse_format(&data).unwrap();
            formats.push((i - 1, fmt));
        }

        let cluster = build_cluster_config(&formats).unwrap();
        assert_eq!(cluster.pools.len(), 1);
        assert_eq!(cluster.pools[0].sets.len(), 1);
        assert_eq!(cluster.pools[0].sets[0].len(), 16);

        // Verify all 16 disks have device_id set
        for disk in &cluster.pools[0].sets[0] {
            assert!(
                disk.device_id.is_some(),
                "disk {} should have device_id",
                disk.uuid
            );
        }
    }
}
